lib/features/login/data/network/config/authorization_interceptors.dart

```diff
@@ -1,32 +1,123 @@
+import 'package:tmail_ui_user/features/login/data/network/oidc_http_client.dart';
 
 class AuthorizationInterceptors extends InterceptorsWrapper {
-  String? _authorization;
+
+  final Dio _dio;
+  final OIDCHttpClient _oidcHttpClient;
+  final TokenOidcCacheManager _tokenOidcCacheManager;
+  final AccountCacheManager _accountCacheManager;
+
+  AuthenticationType _authenticationType = AuthenticationType.none;
+  OIDCConfiguration? _configOIDC;
   Token? _token;
+  String? _authorization;
+
+  AuthorizationInterceptors(
+    this._dio,
+    this._oidcHttpClient,
+    this._tokenOidcCacheManager,
+    this._accountCacheManager
+  );
+
+  void setAuthenticationType(AuthenticationType newAuthenticationType) {
+    _authenticationType = newAuthenticationType;
+  }
+
+  void setOIDCConfiguration(OIDCConfiguration? newConfig) {
+    _configOIDC = newConfig;
+  }
 
   void changeAuthorization(String? userName, String? password) {
     _authorization = base64Encode(utf8.encode('$userName:$password'));
   }
 
   void setToken(Token? newToken) {
     _token = newToken;
+    log('AuthorizationInterceptors::setToken(): newToken: $newToken');
+    log('AuthorizationInterceptors::setToken(): EXPIRE_DATE: ${newToken?.expiredTime?.toIso8601String()}');
   }
 
   @override
   void onRequest(RequestOptions options, RequestInterceptorHandler handler) {
-    if (_authorization != null) {
-      options.headers[HttpHeaders.authorizationHeader] = _getAuthorizationAsBasicHeader(_authorization);
-    } else if (_token != null && _token?.isTokenValid() == true) {
-      options.headers[HttpHeaders.authorizationHeader] = _getTokenAsBearerHeader(_token!.token);
+    switch(_authenticationType) {
+      case AuthenticationType.basic:
+        if (_authorization != null) {
+          options.headers[HttpHeaders.authorizationHeader] = _getAuthorizationAsBasicHeader(_authorization);
+        }
+        break;
+      case AuthenticationType.oidc:
+        if (_token != null && _token?.isTokenValid() == true) {
+          options.headers[HttpHeaders.authorizationHeader] = _getTokenAsBearerHeader(_token!.token);
+        }
+        break;
+      case AuthenticationType.none:
+        break;
     }
     super.onRequest(options, handler);
   }
 
+  @override
+  void onError(DioError err, ErrorInterceptorHandler handler) async {
+    final requestOptions = err.requestOptions;
+    if (_isAuthenticationError(err) && _authenticationType == AuthenticationType.oidc) {
+      if (_configOIDC != null && _token != null) {
+        final newToken = await _oidcHttpClient.refreshingTokensOIDC(
+            _configOIDC!.clientId,
+            _configOIDC!.redirectUrl,
+            _configOIDC!.discoveryUrl,
+            _configOIDC!.scopes,
+            _token!.refreshToken);
+
+        await Future.wait([
+          _tokenOidcCacheManager.persistOneTokenOidc(newToken),
+          _accountCacheManager.setSelectedAccount(Account(
```

Comment:
The `tokenId` is changed when you refresh?. I don't think so

Reply:
Are you sure of that? Just to be on the safe side I want to update it so that an unexpected error never occurs.

Reply:
I am not sure, if `tokenId` not change, we don't need to store account again. We should not keep unneeded account in `Account Cache table`. If `tokenId` change, we should delete the old account.

Reply:
After double checking, I noticed that the `tokenId` value is a string of characters, the same as the `accessToken` value. So after `refreshToken` we always have `tokenId ` changed. So we will do:
- Delete `account` with old `token`.
- Add `account` with new `token`

Reply:
okie, go ahead